p2p/discovery.py

```diff
@@ -820,6 +946,63 @@ def __init__(self, proto: DiscoveryProtocol, peer_pool: PeerPool,
         await self.proto.stop()
 
 
+class NodeTicketInfo:
+    last_issued: int = 0
+    last_used: int = 0
+
+
+class TopicEntry:
+
+    def __init__(self, node: kademlia.Node, expiry: float) -> None:
+        self.node = node
+        self.expiry = expiry
+
+
+class TopicTable:
+    registration_lifetime = 60 * 60
+
+    def __init__(self, logger: TraceLogger) -> None:
+        self.logger = logger
+        # A per-topic FIFO list of nodes.
+        self.topics: Dict[bytes, List[TopicEntry]] = {}
+        # The IDs of the last issued/used tickets for any given node.
+        self.node_tickets: Dict[kademlia.Node, NodeTicketInfo] = {}
+
+    def add_node(self, node: kademlia.Node, topic: bytes) -> None:
+        entries = self.topics.setdefault(topic, [])
+        if len(entries) >= MAX_ENTRIES_PER_TOPIC:
+            entries.pop(0)
+        entries.append(TopicEntry(node, time.time() + self.registration_lifetime))
+
+    def get_nodes(self, topic: bytes) -> Tuple[kademlia.Node, ...]:
+        if topic not in self.topics:
+            return tuple()
+        else:
+            now = time.time()
+            entries = [entry for entry in self.topics[topic] if entry.expiry > now]
+            self.topics[topic] = entries
+            return tuple(entry.node for entry in entries)
+
+    def use_ticket(self, node: kademlia.Node, ticket_serial: int, topic: bytes) -> None:
+        ticket_info = self.node_tickets.get(node)
+        if ticket_info is None:
```

Comment:
Thoughts on using a `try/except` for this and doing the lookup as `ticket_info = self.node_tickets[node]`.  It'll perform worse in the failure case but reduce some overhead in the success case.

Reply:
I think this is more readable and I'm not sure it's worth worrying about that kind of overhead here. Can change if you feel strong about it, though